NE-366: Ingress: Add logLevel and operatorLogLevel configuration for ingress and ingress operator

rfredette opened the pull request:
Implementation of [openshift/enhancements#806](https://github.com/openshift/enhancements/pull/806)

Adds `operatorLogLevel` to ingress.config, and `logLevel` to ingresscontroller, to control the log level of the operator and the log level of the router controller process, respectively.

rfredette commented:
/retest

rfredette commented:
/retest

ahardin-rh commented:
/label docs-approved

CFields651 commented:
/label px-approved

candita commented:
/assign
/assign @miheer

miheer commented on config/v1/types_ingress.go:
```diff
@@ -90,8 +90,28 @@ type IngressSpec struct {
+	// Defaults to "Normal".
+	// +optional
+	// +kubebuilder:default=Normal
+	OperatorLogLevel IngressOperatorLogLevel `json:"operatorLogLevel,omitempty"`
```
Why is it defined in this file ? This file is for config ingress object. Are you going to define log levels in ingress config object ? or in Ingress Controller CR ?

rfredette replied:
Operator log level doesn't need to be defined for each ingress controller, since there's only one instance of the operator for any number of instances of ingress controller resource. I think the ingress config object is the only logical place for it.

rfredette replied:
How would that work if you have mutliple ingress controllers, with each specifying `operatorLogLevel`?

candita replied:
Each has its own pod, its own specification, and its own log.  If I have one shard/controller for prod and one for dev, and I want to log them differently, I would specify the operalorLogLevel on each one's spec, right?

rfredette replied:
There's only one operator instance per cluster, not per ingress controller

candita replied:
I have a hunch that if you remove this from config/v1 it might solve your integration test problem.

deads2k replied:
As others have commented, this isn't the spot you're looking for.  This gets defined one level lower on the operator.openshift.io.  The types in config.openshift.io are logical units, but aren't intended to correspond to individual operators.

rfredette replied:
I'm not sure where else it could go, though, because the only ingress-related API resource in operator.openshift.io is `ingresscontroller`, and that's not for configuration of the operator either. 

Do we need an additional custom resource for operator config? That feels like overkill for just `operatorLogLevel`, but maybe there are other existing/new APIs that make sense there.

miheer commented on config/v1/types_ingress.go:
```diff
@@ -90,8 +90,28 @@ type IngressSpec struct {
 	LoadBalancer LoadBalancer `json:"loadbalancer,omitempty"`
+
+	// logLevel describes the desired logging verbosity for the Ingress Operator
+	// Any one of the following values may be specified:
+	// * Normal
+	// * Debug
```
Vaild values are: "Normal", "Debug".
BTW are you going to add only two levels ?
What log package are you going to use ?

rfredette replied:
The ingress operator uses [zapr](https://github.com/go-logr/zapr) for logging. If the "Normal" level maps to the existing default log level, then Normal is info, and there's only one formal log level more verbose than that in zapr: debug.

That said, it looks like you can add more verbose log levels, they just won't have names, so I'll add one for "Trace" log level

miheer commented on config/v1/types_ingress.go:
```diff
@@ -90,8 +90,28 @@ type IngressSpec struct {
+	// Normal is the default.  Normal, working log information, everything is fine, but helpful notices for auditing or common operations.  In kube, this is probably glog=2.
+	IngressOperatorLogLevelNormal IngressOperatorLogLevel = "Normal"
+
+	// Debug is used when something went wrong.  Even common operations may be logged, and less helpful but more quantity of notices.  In kube, this is probably glog=4.
```
Are we going to set only two log levels ?

candita replied:
The EP shows:
```// Normal is the default.  Normal, working log information, everything is fine, but helpful notices for auditing or common operations.  In kube, this is probably glog=2.
	Normal LogLevel = "Normal"

	// Debug is used when something went wrong.  Even common operations may be logged, and less helpful but more quantity of notices.  In kube, this is probably glog=4.
	Debug LogLevel = "Debug"

	// Trace is used when something went really badly and even more verbose logs are needed.  Logging every function call as part of a common operation, to tracing execution of a query.  In kube, this is probably glog=6.
	Trace LogLevel = "Trace"

	// TraceAll is used when something is broken at the level of API content/decoding.  It will dump complete body content.  If you turn this on in a production cluster
	// prepare from serious performance issues and massive amounts of logs.  In kube, this is probably glog=8.
	TraceAll LogLevel = "TraceAll"
```

candita commented:
[FAIL]  [config.openshift.io/v1, Resource=ingresses] [Stable] Ingress On Create [It] Should be able to create a minimal Ingress
  /go/src/github.com/openshift/api/tests/generator.go:195

/test integration

rfredette commented:
The integration failure looks legitimate; the ingress that was created had `operatorLogLevel: Normal` in the spec. I'll have to address that

candita commented on openapi/openapi.json:
```diff
@@ -7085,6 +7085,10 @@
           "default": {},
           "$ref": "#/definitions/com.github.openshift.api.config.v1.LoadBalancer"
         },
+        "operatorLogLevel": {
```
Why does openapi.json need to change?

rfredette replied:
Yeah, this is auto-generated

candita commented on operator/v1/types_ingress.go:
```diff
@@ -1227,6 +1241,17 @@ type IngressControllerLogging struct {
+
+	// logLevel describes the desired logging verbosity for Router pods.
+	// Any one of the following values may be specified:
+	// * Normal
```
Like in https://github.com/openshift/api/pull/1031/files#diff-46e86efb5c123e387737010ac03828fda0c5edfa402c432480485d928827236cR84-R86 it would be helpful godoc to document what is logged at each of these log levels.

rfredette replied:
I'm not sure I can be as concrete in defining what will be logged at each level, but I will try to give some guidelines/examples

candita commented on operator/v1/types_ingress.go:
```diff
@@ -253,6 +253,20 @@ type IngressControllerSpec struct {
+
+	// Trace is used when something went really badly and even more verbose logs are needed.  Logging every function call as part of a common operation, to tracing execution of a query.  In kube, this is probably glog=6.
+	IngressLogLevelTrace IngressLogLevel = "Trace"
+)
```
The EP also has:
```
	// TraceAll is used when something is broken at the level of API content/decoding.  It will dump complete body content.  If you turn this on in a production cluster
	// prepare from serious performance issues and massive amounts of logs.  In kube, this is probably glog=8.
	TraceAll LogLevel = "TraceAll"
```

rfredette replied:
good catch; I'll add that.

deads2k replied:
This is in the same package we use for our other operators.  Matching the OperatorSpec seems valuable even without embedding the entire type. See https://github.com/openshift/api/blob/master/operator/v1/types.go#L55-L71

candita commented on operator/v1/types_ingress.go:
```diff
@@ -1227,6 +1241,17 @@ type IngressControllerLogging struct {
 	// +optional
 	Access *AccessLogging `json:"access,omitempty"`
+
+	// logLevel describes the desired logging verbosity for Router pods.
```
If this applies to Access logging, it may be better to say:
```suggestion
	// logLevel describes the desired logging verbosity for the Router's Access logs.
```

candita commented on operator/v1/types_ingress.go:
```diff
@@ -1227,6 +1241,17 @@ type IngressControllerLogging struct {
+	// * Normal
+	// * Debug
+	// * Trace
+	//  Setting logLevel: Trace will produce extremely verbose logs.
```
```suggestion
	// Setting logLevel: Trace will produce extremely verbose logs.
```

candita commented on config/v1/types_ingress.go:
```diff
@@ -90,8 +90,31 @@ type IngressSpec struct {
+	// * Debug
+	// * Trace
+	// Defaults to "Normal".
+	// +optional
```
```suggestion
	// +optional
        // +kubebuilder:default=Normal
```

rfredette replied:
I think having the `+kubebuilder:default=Normal` flag in there was previously causing the integration test failure; it was causing `operatorLogLevel: Normal` to get inserted into the otherwise empty default config. That said, after removing the default flag, the description is wrong.

I'm not sure exactly what the common wording is for similar, but I'll update it to something like "When unset, the default log level is 'Normal'"

JoelSpeed replied:
Updating the integration test in this case is ok, adding a new field with a default is not a breaking change. This should be defaulted as it is then consistent with the existing operators which use this field

rfredette commented:
Sorry it took me so long to get back to this, but I've addressed the review comments. @candita and @miheer please take another look when you have a chance

candita commented:
/lgtm

rfredette commented:
Moving back to WIP while I work on a better fit for the `operatorLogLevel` location

JoelSpeed commented on operator/v1/types_ingress.go:
```diff
@@ -1684,3 +1716,69 @@ type IngressControllerList struct {
+// +openshift:compatibility-gen:level=1
+type IngressOperator struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
```
This should be objectmeta, this is what's preventing the generator from running correctly
```suggestion
	metav1.ObjectMeta `json:"metadata,omitempty"`
```

JoelSpeed commented on operator/v1/0000_50_ingress-operator_00-ingressoperator.crd.yaml:
```diff
@@ -0,0 +1,20 @@
+  scope: Namespaced
+  versions:
+    - name: v1
+      schema: {}
```
Just add this extra line else the generator seems to get a little confused
```suggestion
      schema:
        openAPIV3Schema: {}
```

JoelSpeed commented on operator/v1/0000_50_ingress-operator_00-ingressoperator.crd.yaml:
```diff
@@ -0,0 +1,20 @@
+    include.release.openshift.io/ibm-cloud-managed: "true"
+    include.release.openshift.io/self-managed-high-availability: "true"
+    include.release.openshift.io/single-node-developer: "true"
+  name: ingressoperator.operator.openshift.io
```
This should be `<plural>.<group>`
```suggestion
  name: ingressoperators.operator.openshift.io
```

JoelSpeed reviewed (commented):
A few comments to help get the schema generation back on track

rfredette commented:
Since the design now deviates from that in the original enhancement, I've opened a PR to update that enhancement: openshift/enhancements#1314

Assuming that updated design is acceptable, this should be ready for review again. However, I do have a couple open questions. I will add some comments shortly for each of them.

rfredette commented on operator/v1/types_ingress.go:
```diff
@@ -1684,3 +1699,50 @@ type IngressControllerList struct {
+	OperatorLogLevel LogLevel `json:"operatorLogLevel,omitempty"`
+}
+
+type IngressStatus struct{}
```
I added a status field to this new type since it seems like it's expected, but as of right now, it's empty. Would it be better to not define it at all, and add it when necessary?

JoelSpeed replied:
I think it would be good to keep it even though it's empty, we should mark it as a subresource to make sure operators can update it without touching the spec

(We should double check what other operators have in their spec, might find some inspiration)

rfredette commented on operator/v1/types_ingress.go:
```diff
@@ -1684,3 +1699,50 @@ type IngressControllerList struct {
+// Ingress contains configuration options specific to the Ingress Operator itself.
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
```
I added `+openshift:compatibility-gen:level` because it seems to be expected in the root CRDs, but it's not something I think I understand yet, beyond what is said in the comment. Is there a process I should be following to determine the appropriate compatibility level, or a rule of thumb I can use?

JoelSpeed replied:
For a v1 API, level 1 is required and changing it to any other level will cause a validation error. So, not much more to do here right now

JoelSpeed commented on operator/v1/types_ingress.go:
```diff
@@ -1684,3 +1699,50 @@ type IngressControllerList struct {
 }
+
+// +genclient
+// +genclient:nonNamespaced
```
I believe you need both, don't think there's any harm in keeping both so i'd keep both in for now
